codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java

```diff
@@ -265,18 +271,21 @@ public Http2Stream resetSent() {
         }
 
         @Override
-        public final Object setProperty(Object key, Object value) {
-            return data.put(key, value);
+        public final <K, V> V setProperty(PropertyKey<K, V> key, V value) {
+            checkNotNull(key, "key");
+            return properties.add((DefaultPropertyKey) key, value);
```

Comment:
@Scottmitch I'm not sure I agree here...

If the user provides a non-`DefaultPropertyKey`, they'll get a `ClassCastException` which seems appropriate here.

If they provide a `DefaultPropertyKey` from another connection, then we could potentially have a problem, but I think that is a rather silly user error.  We could protect against it by having the key know which connection it belongs to as well, but that may be overkill ... not sure how much we want to worry about this.

Regardless, having the details within the key implementation seems completely reasonable to me and the API seems fairly clear. I can add more documentation if that helps.